src/decorators/withLogger.js

```diff
@@ -0,0 +1,38 @@
+import { set } from '../secret'
+
+import { createDefaultLogger } from '../util/createDefaultLogger'
+import { createConfigError } from '../util/createError'
+import joinAsSpeech, { AND } from '../util/joinAsSpeech'
+import { MISCONFIGURED, MISCONFIGURED_LOGGER } from '../constants'
+import isFunction from '../util/isFunction'
+
+const REQUIRED_METHODS = ['trace', 'debug', 'info', 'warn', 'error', 'fatal']
+
+export function withLogger() {
+  return instance => {
+    if (!instance.configuration) {
+      // Set the logger on the instance
+      set(instance, { logger: createDefaultLogger() })
+      return
+    }
+
+    const logger = instance.configuration.logger || createDefaultLogger()
```

Comment:
Nitpick: What if we did something like this:

```
const { logger = createDefaultLogger() } = instance.configuration
if (!REQUIRED_METHODS.every(method => isFunction(logger[method])) {
	...
}
```

The `createDefaultLogger()` should have all required methods (and is tested) so this should always pass. The only downside I see is that we do the `.every` on the default logger while we shouldn't.

Reply:
yeah that's the mean reason why I did the early return so that I didn't have to to that. Also notice that I check for `if (instance.configuration)` so if we want that then we have to change your code from `instance.configuration` to `instance.configuration || {}`. But I have an idea, hold on.